src/emqx_shared_sub.erl

```diff
@@ -249,58 +256,41 @@ do_pick_subscriber(Group, Topic, round_robin, _ClientId, Count) ->
     Rem + 1.
 
 subscribers(Group, Topic) ->
-    ets:select(?TAB, [{{emqx_shared_subscription, Group, Topic, '$1'}, [], ['$1']}]).
+   [Pid || #emqx_shared_subscriber{subpid = Pid}
+                <- mnesia:dirty_read(?TAB, {Group, Topic})].
+
+groups(Topic) ->
+   lists:usort([Group || #emqx_shared_subscriber{group = Group}
+                <- mnesia:dirty_index_read(?TAB, Topic, #emqx_shared_subscriber.topic)]).
 
 %%-----------------------------------------------------------------------------
 %% gen_server callbacks
 %%-----------------------------------------------------------------------------
 
 init([]) ->
-    {atomic, PMon} = mnesia:transaction(fun init_monitors/0),
-    mnesia:subscribe({table, ?TAB, simple}),
     ets:new(?ALIVE_SUBS, [named_table, {read_concurrency, true}, protected]),
-    {ok, update_stats(#state{pmon = PMon})}.
-
-init_monitors() ->
-    mnesia:foldl(
-      fun(#emqx_shared_subscription{subpid = SubPid}, Mon) ->
-          emqx_pmon:monitor(SubPid, Mon)
-      end, emqx_pmon:new(), ?TAB).
+    {ok, update_stats(#state{})}.
 
 handle_call(Req, _From, State) ->
     emqx_logger:error("[SharedSub] unexpected call: ~p", [Req]),
     {reply, ignored, State}.
 
-handle_cast({monitor, SubPid}, State= #state{pmon = PMon}) ->
-    NewPmon = emqx_pmon:monitor(SubPid, PMon),
+handle_cast({subscribe, SubPid}, State= #state{}) ->
     ok = maybe_insert_alive_tab(SubPid),
-    {noreply, update_stats(State#state{pmon = NewPmon})};
```

Comment:
I am not quite following why there is no need to monitor any more.
In case it is a remote pid (i.e. from another node), if the node is down, (or connection is cut),
the remote subscriber may not have a chance to unsubscribe.

Even for a local subscriber, it could be killed, i.e. no chance to evaluate `gen_server:terminate/2`